proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+# Cryptography Community Project: Leg 1 Proposal
+
+# Abstract
+
+This community project proposal is for full-time funding for the development of a suite of Haskell cryptography libraries and tooling suitable a wide range of uses including data integrity, privacy, security, and networking. This is necessary to relieve the burden [of correctly implementing cryptography] from Haskell developers seeking to provide privacy and security to their users. We propose the development of a community-owned set of cryptography libraries of increasing capability, starting with bindings to a compatible open-source cryptography library. This is expected to require a considerable effort over time, and if accepted will be broken up into several legs, each of which will be accompanied by its own specific proposal and set of goals.
+
+The first leg of this proposal is for the development of high-quality, production-ready bindings to the Botan C++ cryptography library, and from acceptance to release on Hackage is estimated to take 3 months of full-time development. We see several challenges but no significant technical risk to this proposal; known challenges involve strictness and developing additional bindings to C++; these items are of high-value and may take significant effort to resolve, but are known to be solvable. Considerable progress has already been made, and the proposed funding will help ensure its completion.
+
+# Background
+
+Cryptography is an essential part of modern computing and information security, but it requires care, nuance, and knowledge to properly implement, and is extremely easy to mess up. It covers a diverse range of functions, with individual algorithms often having their own particular quirks that must be acknowledged to properly and safely use. Failure to properly implement cryptography carries the risk of severely compromising the security of any system in which it is used. Manually implementing common cryptographic primitives such as hashes, ciphers, digital signatures, big-integer arithmetic is thus considered extremely unwise, both in terms of security and performance.
+
+Cryptography is even harder for functional programming languages. Properly implementing techniques such as sanitizing memory, performing operations in constant time, are already difficult enough in languages such as C, and in a lazy functional language it is made even more difficult by such things as the potential for references to be kept alive within thunks. As a result, the space of functional cryptography is relatively under-explored and under-developed, having favored other languages where such things are easier to manage.
```

Comment:
Sorry, That is just colloquialism / lax terminology. I am not sure whether thunk vs closure matters - the concern is that secure data may be captured, whether itself as part of a closure, or because its cleanup was be deferred in a thunk and not forced, and either way it can result in sensitive data staying in memory for longer than intended.

Lazy evaluation presents a challenge towards managing and clearing sensitive data\* from memory, as care must be taken in order to force finalizers / cleanup appropriately. This is in addition to the challenges that strict languages already face (eg, ejecting from caches / registers).

> \* keys, secure texts, personally identifiable information

I should probably also emphasize that this effect pertains to **lazy evaluation**, not *functional languages in general*.